lib/trento/application/process_managers/deregistration_process_manager.ex

```diff
@@ -84,12 +138,147 @@ defmodule Trento.DeregistrationProcessManager do
+          host_id: host_id,
+          deregistered_at: requested_at
+        }
+      end),
+      %DeregisterHost{host_id: host_id, deregistered_at: requested_at}
+    ])
+  end
+
+  def apply(
+        %DeregistrationProcessManager{database_instances: database_instances} = state,
+        %DatabaseInstanceRegistered{
+          sap_system_id: sap_system_id,
+          instance_number: instance_number
+        }
+      ) do
+    %DeregistrationProcessManager{
+      state
+      | database_instances: [
+          %Instance{
+            sap_system_id: sap_system_id,
+            instance_number: instance_number
+          }
+          | database_instances
+        ]
+    }
+  end
+
+  def apply(
+        %DeregistrationProcessManager{application_instances: application_instances} = state,
+        %ApplicationInstanceRegistered{
+          sap_system_id: sap_system_id,
+          instance_number: instance_number
+        }
+      ) do
+    %DeregistrationProcessManager{
+      state
+      | application_instances: [
+          %Instance{
+            sap_system_id: sap_system_id,
+            instance_number: instance_number
+          }
+          | application_instances
+        ]
+    }
+  end
+
+  def apply(
+        %DeregistrationProcessManager{database_instances: database_instances} = state,
+        %DatabaseInstanceDeregistered{instance_number: instance_number}
+      ) do
+    %DeregistrationProcessManager{
+      state
+      | database_instances:
+          Enum.reject(database_instances, fn %Instance{
+                                               instance_number: current_instance_number
+                                             } ->
+            current_instance_number == instance_number
+          end)
+    }
+  end
+
+  def apply(
+        %DeregistrationProcessManager{application_instances: application_instances} = state,
+        %ApplicationInstanceDeregistered{instance_number: instance_number}
+      ) do
+    %DeregistrationProcessManager{
+      state
+      | application_instances:
+          Enum.reject(application_instances, fn %Instance{
+                                                  instance_number: current_instance_number
+                                                } ->
+            current_instance_number == instance_number
+          end)
+    }
+  end
+
   def apply(%DeregistrationProcessManager{} = state, %ClusterRolledUp{
         cluster_id: cluster_id
       }) do
     %DeregistrationProcessManager{state | cluster_id: cluster_id}
   end
 
+  def apply(
+        %DeregistrationProcessManager{} = state,
+        %SapSystemRolledUp{
+          sap_system_id: sap_system_id,
+          snapshot: %SapSystem{
+            database: %SapSystem.Database{instances: database_instances},
+            application: %SapSystem.Application{instances: application_instances}
+          }
+        }
+      ) do
+    %DeregistrationProcessManager{
+      state
+      | database_instances:
```

Comment:
This logic looks wrong. You are replacing completely the database and application instances, but be aware that one host can have applications/databases from different sap systems.
So you would need to get other instances from other sap systems and concat with the ones coming in the rollup